software/contrib/bernoulli_gates.py

```diff
@@ -0,0 +1,218 @@
+from europi import *
+from time import sleep_ms
+from random import random
+from europi_script import EuroPiScript
+import machine
+
+# Constant values for display
+BERNOULLI_BAR_LEN = 50
+BERNOULLI_BAR_WID = 6
+BERNOULLI_INDI_WID = 5
+
+# Trigger Time (some module may not be able to catch up the trigger if it is too short)
+BERNOULLI_TRG_T = 20
+
+class SingleBernoulliGate():
+    def __init__(self, control_knob =  k1, control_port = ain, out_port = (cv1, cv2, cv3), visualization_para = ('P1', (0, 0), 0, -3), port3_func = 'clock', port3_source_cv = None):
+        '''
+        1. control_knob and control_port tunes the probability threshold
+        if control_port == None, probability threshold is only determined by control_knob
+        2. out_port assigns left, right, and function port accordingly
+        out_port shoud have at least 2 ports (port3_func == 'none')
+        3. visualization_para = (text, text_pos, tex1_pos, bar_pos_offset)
+        4. port3_func can be 'none', 'clock', 'and', 'or', 'xor'
+        the output of function_port = logic(current left, source), logic can be and, or, xor
+        '''
+        self.mode_flg = 0 # mode 0: Trigger, mode 1: Gate, mode 2: Toggle
+        self.coin = 0 # probability
+        self.right_possibility = 0 # change every iteration
+        self.left_possibility = 0
+        self.right_possibility_sampled = 0 # only change when triggered
+        self.left_possibility_sampled = 0
+
+        self.port3_func = port3_func
+        self.port3_source_cv = port3_source_cv
+
+        self.control_knob = control_knob
+        self.control_port = control_port
+
+        self.left_port = out_port[0]
+        self.right_port = out_port[1]
+        if port3_func != 'none':
+            self.function_port = out_port[2]
+
+        self.text = visualization_para[0]
+        self.text_pos = visualization_para[1]
+        self.text1_pos = visualization_para[2]
+        self.bar_pos_offset = visualization_para[3]
+
+    def get_prob(self):
```

Comment:
Typically methods whose names start with `get` return values. A name like `update_prob` might be more clear.